src/Biogeochemistry.jl

```diff
@@ -0,0 +1,197 @@
+"""
+update_biogeochemical_state!(bgc, model) = nothing
+
+@inline biogeochemical_drift_velocity(bgc, val_tracer_name) = nothing
+@inline biogeochemical_advection_scheme(bgc, val_tracer_name) = nothing
+@inline biogeochemical_auxiliary_fieilds(bgc) = NamedTuple()
+
+#####
+##### Default (discrete form) biogeochemical source
+#####
+
+abstract type AbstractBiogeochemistry end
+
+@inline function biogeochemistry_rhs(i, j, k, grid, bgc, val_tracer_name::Val{tracer_name}, clock, fields) where tracer_name
+    U_drift = biogeochemical_drift_velocity(bgc, val_tracer_name)
+    scheme = biogeochemical_advection_scheme(bgc, val_tracer_name)
+    src = biogeochemical_transition(i, j, k, grid, bgc, val_tracer_name, clock, fields)
+    c = @inbounds fields[tracer_name]
+        
+    return src - div_Uc(i, j, k, grid, scheme, U_drift, c)
+end
+
+@inline biogeochemical_transition(i, j, k, grid, bgc, val_tracer_name, clock, fields) =
+    bgc(i, j, k, grid, val_tracer_name, clock, fields)
+
+@inline (bgc::AbstractBiogeochemistry)(i, j, k, grid, val_tracer_name, clock, fields) = zero(grid)
+
+#####
+##### Continuous form biogeochemical source
+#####
+ 
+"""Return the biogeochemical forcing for `val_tracer_name` when model is called."""
+abstract type AbstractContinuousFormBiogeochemistry <: AbstractBiogeochemistry end
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{1}) =
+    @inbounds (fields[names[1]][i, j, k],)
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{2}) =
+    @inbounds (fields[names[1]][i, j, k],
+               fields[names[2]][i, j, k])
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{N}) where N =
+    @inbounds ntuple(n -> fields[names[n]][i, j, k], Val(N))
+
+@inline function biogeochemical_transition(i, j, k, grid, bgc::AbstractContinuousFormBiogeochemistry,
+                                           val_tracer_name, clock, fields)
+
+    names_to_extract = tuple(required_biogeochemical_tracers(bgc)...,
+                             required_biogeochemical_auxiliary_fields(bgc)...)
+
+    fields_ijk = extract_biogeochemical_fields(i, j, k, grid, fields, names_to_extract)
+
+    x = xnode(Center(), Center(), Center(), i, j, k, grid)
+    y = ynode(Center(), Center(), Center(), i, j, k, grid)
+    z = znode(Center(), Center(), Center(), i, j, k, grid)
+
+    return bgc(val_tracer_name, x, y, z, clock.time, fields_ijk...)
+end
+
+@inline (bgc::AbstractContinuousFormBiogeochemistry)(val_tracer_name, x, y, z, t, fields...) = zero(x)
+
+struct NoBiogeochemistry <: AbstractBiogeochemistry end
+
+tracernames(tracers) = keys(tracers)
+tracernames(tracers::Tuple) = tracers
+
+@inline function has_biogeochemical_tracers(fields::NamedTuple, required_fields, grid)
+
+    for field_name in required_fields
+        if !(field_name in keys(fields))
+            fields = merge(fields, NamedTuple{(field_name, )}((CenterField(grid), )))
+        end
+    end
```

Comment:
@jagoosw this is hard to parse. Does this add a `CenterField` when one of the required BGC fields isn't found in a named tuple of tracers (eg because the tracers have been predefined by the user?)

I feel this might lead to subtle bugs in user code, ie when they mispell a name when building a named tuple, the user won't get an error. Instead the model will run but will silently be wrong / incorrect in a suble way that might be hard to figure out.

I suggest for the case of `NamedTuple` we throw a helpful error if all the fields are not provided ("The names x, y, z were not found but need to be included with <biogeochemical model>..." or something.

Reply:
So that a user can add their own tracers, and we automatically add tracers, we probably need to check if all of the bgc tracers are present - or if none of the bgc tracers are present (in which case we add them) - otherwise we error?

Reply:
So:
```
user_set_all_tracers = all(tracer ∈ tracers for tracer in req_tracers)
user_set_some_tracers = any(tracer ∈ tracers for tracer in req_tracers)

if user_set_some_tracers && !user_set_all_tracers
    error("useful error saying that the model uses those tracer names, or that they need to define all of the names")
elseif !user_set_all_tracers
    for ... # as above end
end
```
?

Reply:
```julia
if !all(name in req_tracers for name in keys(tracers))
    throw(ArgumentError("Helpful error about required tracer fields in tracers NamedTuple"))
end

return nothing
```